[RDY] Refactor/enhance job api

tarruda opened the pull request:
This is the refactor I mentioned in #2076. It aims to simplify and optimize the job API, removing the `JobActivity`/`v:job_data` and calling functions instead of autocmds.

It is also necessary to avoid deferred events and needless preprocessing of job data when a terminal is spawned(Currently every chunk of data received from the job is split into a list of lines with NULs replaced by linefeeds, even if no job handlers are registered)

Close #1838

tarruda committed:
eval: Refactor `call_func` and `func_unref`

- Make it possible to call or unref ufunc_T pointers directly.
- Keep refcount of named functions, and stop them from being deleted if the
  refcount is greater than 1.

tarruda committed:
eval: Refactor vimscript job control API

- Remove JobActivity autocmd and v:job_data variable
- Simplify `jobstart` to receive:
  - An argument vector
  - An optional dictionary which may contain any of the current `jobstart`
    options plus `on_stdout`, `on_stderr` and `on_exit` callbacks.
- Refactor and add more job tests
- Update documentation

rainerborene commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
+jobstart({argv}[, {opts}])			{Nvim} *jobstart()*
+		Spawns {argv}(list) as a job. If passed, {opts} must be a
+		dictionary with the any of the following keys:
+		  - on_stdout: stdout event handler
```
I suggest you to use `on` prefix without underscores just like the HTML spec, also note that none of the vim functions have underscores although being dictionary keys here.

rainerborene commented:
@tarruda any reason to use `on` prefix with underscores? why not follow the HTML spec on that matter?

justinmk commented:
@rainerborene 

- our style guide suggests underscores for C functions
- it would be unpleasant to name an event handler `onevent_occurred` instead of `on_event_occurred`, for example. 
- DOM event names probably were named as such because they began life as inline attributes (e.g. `<td onclick="...">`), before

bfredl commented:
I don't remember if it was discussed before or not, but have you considered using/supporting the already existing `Dictionary-Function` pattern, i e when a function is defined directly inside a dict, it can access that dict through an implicit `self` argument:
```
let mydict = {'data': [0, 1, 2, 3]}
function mydict.len()
   return len(self.data)
endfunction
```

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
-		|JobActivity| events. It returns:
+jobstart({argv}[, {opts}])			{Nvim} *jobstart()*
+		Spawns {argv}(list) as a job. If passed, {opts} must be a
+		dictionary with the any of the following keys:
```
typo: with the any

tarruda replied:
:+1:

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
+		  - TERM: $TERM environment variable(only if pty is set)
+		Either funcrefs or function names can be passed as event
+		handlers.
+		Returns:
```
`{prog}` is still used in the description of return value `-1` some lines below.

tarruda replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
+		Either funcrefs or function names can be passed as event
+		handlers.
+		Returns:
 		  - The job id on success, which is used by |jobsend()| and
```
should probably be `ID`

tarruda replied:
:+1:

oni-link commented on runtime/doc/job_control.txt:
```diff
@@ -82,15 +74,19 @@ Here's what is happening:
+  2: User data set in the "user" option passed to |jobstart()|. Can be any
+     vimscript object and is used to identify the context when multiple
+     |jobstart()| calls share a single set of callbacks.
+  3:  is "stdout" or "stderr", this will contain a list of
```
"exit" is missing. See `1:`.

tarruda replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
+		  - pty: If set, the job will be connected to a new pseudo
+		    terminal, and the job streams are connected to the master
+		    file descriptor.
+		  - width: Width of the terminal screen(only if pty is set)
```
```
`pty` might be better (with the accents around it), to clarify that you're talking about the argument and not a pseudo terminal.
```

tarruda replied:
The text above makes it clear that what follows are dictionary options, I don't think its possible to confuse the reader here.

Pyrohh replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -4024,10 +4024,22 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
+		  - on_exit: exit event handler
+		  - user: user data passed to all callbacks
+		  - pty: If set, the job will be connected to a new pseudo
+		    terminal, and the job streams are connected to the master
```
`are connected` -> `will be connected`, unless this is on purpose

tarruda replied:
:+1:

Pyrohh commented on runtime/doc/job_control.txt:
```diff
@@ -37,37 +37,29 @@ for details
-|jobsend()| and |jobstop()| functions, and by listening to the |JobActivity|
-event. The best way to understand is with a complete example:
+|jobsend()| and |jobstop()| functions. The best way to understand is with a
+complete example:
```
maybe just `Here's an example:`

tarruda replied:
:+1:

Pyrohh commented on runtime/doc/job_control.txt:
```diff
@@ -37,37 +37,29 @@ for details
+    \ 'on_exit': function('s:JobHandler')
+    \ }
+    let job1 = jobstart(['bash'], extend({'user': 'shell 1'}, s:callbacks))
+    let job2 = jobstart(['bash', '-c', 'for ((i = 0; i < 10; i++)); do echo hello $i!; sleep 1; done'], extend({'user': 'shell 2'}, s:callbacks))
```
Since we're using bash, then why not `for i in {1..10}`?

tarruda replied:
:+1: 

I wasn't familiar with the range syntax, thanks

oni-link commented on src/nvim/eval.c:
```diff
@@ -10717,55 +10727,62 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv)
+  char **argv = xcalloc(argc + 1, sizeof(char *));
+  for (listitem_T *arg = args->lv_first; arg != NULL; arg = arg->li_next) {
+    argv[i++] = xstrdup((char *)arg->li_tv.vval.v_string);
+  }
```
`argv[i] = NULL;` is missing.

splinterofchaos replied:
`v_string` may be `NULL` in some circumstances to represent an empty string. `get_tv_string()` would return an empty string in this case, and also allow the implicit conversion of integers to strings. Same comment for the `os_can_exe()` call above.

ref: #1471 for an instance where this may segfault.

oni-link commented on src/nvim/eval.c:
```diff
@@ -19829,6 +19893,25 @@ static inline JobOptions common_job_options(char **argv, char *autocmd_file)
+  *on_stderr = get_dict_callback(vopts, "on_stderr");
+  *on_exit = get_dict_callback(vopts, "on_exit");
+  if (did_emsg) {
+    return;
```
Could need `user_func_unref(on_*);` calls.

tarruda replied:
:+1:

tarruda commented:
> I don't remember if it was discussed before or not, but have you considered using/supporting the already existing Dictionary-Function pattern, i e when a function is defined directly inside a dict, it can access that dict through an implicit self argument:

Is there a difference from the "user" option? Its already possible to do this:

```vim
function s:Handler(id, data, user)
endfunction

call jobstart([..], {'on_stdout': 's:Handler', 'user': {}})
```

bfredl commented:
Nothing except consistency with existing  vimscript functionality. Not that
I know if theyre actaully used to any extent, but if we're settling for
funcrefs within a dict as callback pattern, there should be an explicit
desicion to either support or deprecate dict-functions.
Den 27 mar 2015 01:01 skrev "Thiago de Arruda" <notifications@github.com>:

> I don't remember if it was discussed before or not, but have you
> considered using/supporting the already existing Dictionary-Function
> pattern, i e when a function is defined directly inside a dict, it can
> access that dict through an implicit self argument:
>
> Is there a difference from the "user" option? Its already possible to do
> this:
>
> function s:Handler(id, data, user)endfunction
>
> call jobstart([..], {'on_stdout': 's:Handler', 'user': {}})
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/neovim/neovim/pull/2247#issuecomment-86759805>.
>

tarruda commented:
> Nothing except consistency with existing  vimscript functionality. Not that
I know if theyre actaully used to any extent, but if we're settling for
funcrefs within a dict as callback pattern, there should be an explicit
desicion to either support or deprecate dict-functions.

I could change the "user" option to "self" and pass it as a implicit self argument, but nothing would stop the user from doing this:

```vim
function s:Handler(id, data)
  self.id = 1  " self is available
endfunction

call jobstart([..], {'on_stdout': 's:Handler', 'self': {id: 0}})
```

Even though s:Handler is not a dictionary function. Also, I don't think there's a way avoid having to pass the "self" function despite using a dictionary function as callback:

```vim
let dict = {id: 1}
function dict.handler(id, data)
  echo self.id  " self.id == 0
endfunction

call jobstart([..], {'on_stdout': dict.handler, 'self': {id: 0}})
```

That is, self will always be what is passed to `jobstart`, not the dict where the function was defined(unless the vimscript engine "binds" dict.handler to dict, but I don't think thats the case)

pgdouyon commented:
Sorry if I'm misunderstanding something and this went completely over my head, but I understood @bfredl's suggestion as something like this:

```vim
let dict = {'on_stdout': function("s:Handler"), 'id': 0}
function! s:Handler() dict
    echo self.id
endfunction

call jobstart([..], dict)
```

Where `self` is the dictionary passed to jobstart, not a key in that dictionary, so that handler parameters can be stored directly in the dictionary and not passed through the "user" entry.  Personally, I think calling the handlers as dictionary functions is a bit more elegant, but the "user" key is valuable in that it provides a separate namespace for parameters and avoids collisions with job-control options.

kopischke commented:
@bfredl

> I know if theyre actaully used to any extent, but if we're settling for funcrefs within a dict as callback pattern, there should be an explicit desicion to either support or deprecate dict-functions.

I use dictionary functions all the time in my plugins – they are one of the rare really high level Vin script features, allowing OOP type encapsulation of behaviour.

@tarruda

> self will always be what is passed to `jobstart`, not the dict where the function was defined (unless the vimscript engine "binds" dict.handler to dict, but I don't think thats the case)

It doesn't – the behaviour you describe is consistent with what `function()` does: it binds its optional third argument (which needs to be a Dictionary) to `self` for the purpose of a dictionary Funcref passed as its first argument. It’s not a perfect solution by a far cry (what in Vim script is?), but I am not sure adding yet another set of Funcref semantics is going to help with that :).

bfredl commented:
@tarruda I didn't mean introduce yet another semantics for funcref passing, just supporting the _existing_ semantics for calling a function inside a dict, i e if jobstart "were a vimscript function" the following would be expected to work
```
let handlers = {'myid':'0'}
function handlers.on_stdout(id, data)
    "here self.myid is available
end

function handlers.on_exit(id, data)
    "also here
end

call jobstart(args, handlers)

" separate instance
call jobstart(args, extend(handlers, {'myid':1}))
```
but as @pgdouyon  pointed out, there would be a namespace conflict with the options here, so if this pattern were "actively" supported, there would probably be a separate "handlers" argument/option.

tarruda commented:
I've refactored to be more friendly with the dictionary pattern. Even with the possible namespace clashes, I think its simpler to reuse the job options object as "self", its not like the user will run out of keys :smile:

Pyrohh commented on runtime/doc/job_control.txt:
```diff
@@ -37,37 +37,28 @@ for details
+    \ 'on_exit': function('s:JobHandler')
+    \ }
+    let job1 = jobstart(['bash'], extend({'shell': 'shell 1'}, s:callbacks))
+    let job2 = jobstart(['bash', '-c', 'for ((i = 0; i < 10; i++)); do echo hello $i!; sleep 1; done'], extend({'shell': 'shell 2'}, s:callbacks))
```
`{1..10}` could be used here too, I must have missed it the first time.

tarruda replied:
:+1:

tarruda commented:
Added some basic documentation for the terminal emulator. Didn't elaborate much because some things will be changing in the next days as we get feedback from users(#2171, #2257 ...)

Pyrohh commented:
@tarruda I sent you a PR with some improvements.

tarruda commented:
> @tarruda I sent you a PR with some improvements (no elaboration as you said, just low hanging fruit).

:+1: applied your fixes, thanks

tarruda committed:
eval: Implement `jobwait()` vimscript function

tarruda committed:
eval: Implement `jobclose()` vimscript function

tarruda commented:
Implemented the `jobwait` function which should allow @junegunn to close junegunn/vim-plug#104

Also modified `jobsend` so that the second argument is optional. If omitted, the job stdin is closed. @bfredl can you verify if this meets #1838 requirements?

tarruda committed:
eval: Improve validation of ids passed to job functions

Use the `is_user_job` to ensure that the job was started by `jobstart` or
`termopen`.

tarruda committed:
doc: Begin terminal emulator documentation

With some spacing/indentation fixes.

Helped by: @Pyrohh, @kopischke

bfredl commented:
@tarruda It does what's required, but enables to crash nvim:
```
let j = jobopen(["xsel", "-n", "-i"])
call jobsend(j, ["some text"])
call jobsend(j)
call jobsend(j, ["this crashes nvim"])
```
Also, would it hurt to have separate `jobclose` ? ( `jobsend(j, data); jobsend(j)` isn't really self-explanatory)

tarruda commented:
@bfredl Implemented your `jobclose` suggestion

ZyX-I commented:
> I could change the "user" option to "self" and pass it as a implicit self argument, but nothing would stop the user from doing this:
>
>     function s:Handler(id, data)
>       self.id = 1  " self is available
>     endfunction
>
>     call jobstart([..], {'on_stdout': 's:Handler', 'self': {id: 0}})

If you did not alter code that calls user functions then `self` is *not* passed to non-dictionary function even if it was given explicitly by `call`. Try the following:

```VimL
function NonDict()
    echo l:
endfunction
function Dict() dict
    echo l:
endfunction
call call('NonDict', [], {}) " echoes {}
call call('Dict', [], {}) " echoes {'self': {}}
```

bfredl commented:
Just noted a small discrepancy between `jobstart()` and `termopen()`, while the former accepts an argv-list only, the later only accepts a shell command string. Any particular reason for this or wouldn't it make more sense for both functions to accept both formats?

tarruda commented:
> Just noted a small discrepancy between jobstart() and termopen(), while the former accepts an argv-list only, the later only accepts a shell command string. Any particular reason for this or wouldn't it make more sense for both functions to accept both formats?

No reason, it's simply a consequence of how `termopen()` was extracted from `:terminal`. For `:terminal` I think its important to continue using the shell so pipes and other shell features can be used naturally in nvim command line, but I `termopen` should probably be refactored to receive `argv` like `jobstart`.

justinmk commented:
> any reason to use on prefix with underscores? why not follow the HTML spec on that matter?

@rainerborene I owe you a retraction: I realize now you were talking about the vimscript `"on_stdin"`, `"on_stdout"`, `"on_stderr"` keys, not the C source code. Yes, I agree we should avoid underscores in vimscript identifiers, if possible.

justinmk commented:
> I think its important to continue using the shell so pipes and other shell features can be used naturally in nvim command line, but termopen should probably be refactored to receive argv like jobstart.

:+1:
